### DISCUSSION TITLE
Fetch and store data in Pinia store before Nuxt application renders

### DISCUSSION THREAD

**Discussion by Zibbp** [nuxt/nuxt#17315] [Questions]
Hi,
I'm having trouble finding where and how to execute some code to fetch user data from an API using a cookie and store the data in a Pinia store. Currently I'm using an `onMounted` hook in `app.vue`. I believe this is not firing soon enough as I have middleware which logs the store as empty.

```
onMounted(async () => {
  await authStore.init();
});
```

An example middleware fetching state from the store.

```
import { useAuthStore } from "~/stores/AuthStore";

export default defineNuxtRouteMiddleware((to, from) => {
    const authStore = useAuthStore()
    console.log(authStore.user.role)
})
```

The middleware logs it as null, but I do see the state as being populated using the Vue devtools. This leads me to believe the `onMounted` in `app.vue` is running too late.

Is there a more appropriate place to put some code that will allow me to update a Pinia store and that needs to run before the application is rendered and middleware is used?

**Comment by milelazar** [comment]
You can create a plugin and from there you can fetch the data and populate the store before app is mounted. It works for me in a similar scenario. Here is an example

plugins/index.js
```
import { useAuthStore } from "~/stores/AuthStore";

export default defineNuxtPlugin(async (nuxtApp) => {
	const authStore = useAuthStore()
	await authStore.init()

	addRouteMiddleware(async (to, from) => {
		// you can have middleware here if needed
	})
})
```

You should have store data available once app is loaded. Hope this helps
